Enforce strict id:alias slug validation for asset downloads

shinde-rahul opened the pull request:
It is necessary to know more than just the ID to download an asset

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | ❌ <!-- Use emojis to indicate positive (green) or negative (red) for each item in the table. -->
| New feature/enhancement? (use the a.x branch)      | ✔️
| Deprecations?                          | ❌
| BC breaks? (use the c.x branch)        | ❌
| Automated tests included?              | ✔️ <!-- All PRs must maintain or improve code coverage -->
| Related user documentation PR URL      | mautic/user-documentation#... <!-- required for new features -->
| Related developer documentation PR URL | mautic/developer-documentation-new#... <!-- required for developer-facing changes -->
| Issue(s) addressed                     | Fixes #... <!-- prefix each issue number with "Fixes #", no need to create an issue if none exists, explain below instead -->

## Description

This PR strengthens asset download security in Mautic by enforcing **strict validation of the asset slug** used in public download URLs.

### Problem:

Previously, if someone guessed an asset ID, they could download it directly using a partial or malformed slug.
For example:
https://mautic.ddev.site/asset/1: would still return a valid file.

This exposes all assets to unauthorized access via ID enumeration.

### Solution:

The download mechanism has been refactored to:

* Require **strict slug format**: `id:alias`
* Reject partial or malformed slugs with a proper `404 Not Found`
* Prioritize alias for canonical URLs, improving stability and preventing guesswork
* Track unauthorized or redirected access attempts
* Streamline response logic (e.g., published, remote/local asset, robots headers)

All of this is done via:

* A new repository method: `findByIdAndAlias`
* A rewritten and test-covered `downloadAction` in `PublicController`

###  Additional Improvements

* Canonical download URLs now use `id:alias`
* Functional tests added for:
  * Invalid or malformed slugs
  * Redirect behavior
  * Unauthorized access to unpublished assets
  * Remote and missing asset handling
  * Proper robot headers for disallowed downloads

### Examples:

| URL                                             | Expected Result                 |
| ----------------------------------------------- | ------------------------------- |
| `/asset/53:my-alias`                            | ✅ File steams                |
| `/asset/53:`                                    | ❌ Returns 404                   |
| `/asset/53`                                     | ❌ Returns 404                   |
| `/asset/53:wrong-alias`                         | ❌ Returns 404                   |
| Unpublished asset                               | 🔒 403 Forbidden                |


---
### 📋 Steps to test this PR:

1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. Create a new asset via the UI.
3. Note the asset ID and alias.
4. Try downloading the asset using different URLs:
    1. `https://mautic.ddev.site/asset/<ID>:<alias>` → ✅ File should stream or download.
    2. `https://mautic.ddev.site/asset/<ID>:<wrong-alias>` → ❌ Should return 404.
    3. `https://mautic.ddev.site/asset/<ID>:` → ❌ Should return 404.
    4. `https://mautic.ddev.site/asset/<ID>` → ❌ Should return 404.
6. Confirm that only complete, valid slugs work for downloads.

volha-pivavarchyk committed:
not possible to download assets using only IDs

volha-pivavarchyk committed:
fix the redurect to another route when the asset ID is used

volha-pivavarchyk committed:
fix phpstan issues

volha-pivavarchyk committed:
add test the case when an asset is downloaded by id

Bastian2718 reviewed (approved):
Looks good to me!

jos0405 reviewed (approved):
<img width="1012" height="610" alt="Screenshot 2025-07-10 at 18 05 02" src="https://github.com/user-attachments/assets/f1fcf2d6-5c3b-4e34-b729-605f7e2a924b" />

I'm unable to download the asset, so I can no longer 'guess' the path.

shinde-rahul committed:
Merge branch '7.x' into feature/not-possible-to-download-assets-using-only-IDs

shinde-rahul committed:
Refactor: Simplify PublicController::downloadAction to reduce complexity

shinde-rahul committed:
Enhanced test coverage for \Mautic\AssetBundle\Controller\PublicController

shinde-rahul committed:
Covered `X-Robots-Tag` header for disallowed assets scenario as well

shinde-rahul committed:
Updated exception to use the dedicated exception than generic

shinde-rahul commented:
@RCheesley, I have refactored and added missing coverage. The only issue reported by SonarQube is a false positive.

@jos0405 and @Bastian2718, please re-review the PR.

RCheesley reviewed (approved):
Thanks for the work on this @shinde-rahul - confirm that it's working as expected and I'm not able to access local or remote URLs without providing the slug in addition to the ID of the asset.

RCheesley committed:
Merge branch '7.x' into feature/not-possible-to-download-assets-using-only-IDs

matbcvo commented:
Wouldn't using UUIDs be a better option here (instead of ID+alias)? UUIDs are unguessable and eliminates the risk of ID enumeration.

kuzmany reviewed (changes requested):
If I type with fake lias:

.../asset/1:test

url is redirect to  real asset url

../asset/1:thumbnailprzklinec20251200x63023jpg

I assume  fake alias should return 404 error too.

shinde-rahul committed:
refactor(asset): use UUID-based canonical slugs, simplify resolution, and improve redirect accuracy

- Removed regex-based slug validation in PublicController and replaced with repository-based resolution using findByIdAndAliasOrUuid().
- Updated AssetRepository to add the new method to locate assets via ID and either alias or UUID, improving backward compatibility and robustness.
- Canonical URLs are now generated using UUID instead of alias, ensuring consistency and preventing partial or malformed slug access.
- Updated AssetModel::generateUrl() to include optional 'stream' parameter and fallback to UUID over alias for slug generation.
- Fixed data-copy download link in asset listing template to use UUID-based slug.
- Updated functional tests to expect UUIDs in slugs and fixed redirect behavior assertions.
- Improved redirect response to preserve stream parameters.
- Fixed ExceptionController to safely fallback when 'path' is missing in parsed URLs.

This commit reduces cognitive complexity in the controller, improves URL handling consistency, and prepares the system for future-safe, UUID-first identifiers.

shinde-rahul committed:
Merge branch 'feature/not-possible-to-download-assets-using-only-IDs' of github.com:shinde-rahul/mautic into feature/not-possible-to-download-assets-using-only-IDs

shinde-rahul committed:
Merge branch '7.x' into feature/not-possible-to-download-assets-using-only-IDs

shinde-rahul committed:
CS fixed

shinde-rahul committed:
Fixing asset test cases

shinde-rahul committed:
Updated the code to keep alias only and updated tests accordingly

shinde-rahul commented on app/bundles/AssetBundle/Resources/views/Asset/list.html.twig:
```diff
@@ -180,7 +180,7 @@
                                     1: {
                                         'attr' : {
-                                            'data-copy' : url('mautic_asset_download', {'slug': item.getId() ~ ':' ~ item.getAlias()}),
+                                            'data-copy' : url('mautic_asset_download', {'slug': item.getId() ~ ':' ~ item.getUuid()}),
```
```suggestion
                                            'data-copy' : url('mautic_asset_download', {'slug': item.getId() ~ ':' ~ item.getAlias()}),
```

shinde-rahul committed:
Update app/bundles/AssetBundle/Resources/views/Asset/list.html.twig

shinde-rahul committed:
Merge branch '7.x' into feature/not-possible-to-download-assets-using-only-IDs

shinde-rahul commented:
@kuzmany – I've updated the code to validate both the ID and the alias in the asset slug as discussed.

@matbcvo – I initially considered using the UUID in the URL, but after discussing it with @adiux, we decided against it. UUIDs might appear suspicious or "malicious-looking" to users, which could reduce trust and click-through rates.

@RCheesley @jos0405 @kuzmany @matbcvo and @Bastian2718 – The implementation has now been revised. I'd appreciate it if you could take another look and re-review the changes when you get a chance.

Thanks! 🙌

kuzmany commented:
@shinde-rahul 

For me, I would take this as good BC and UUID  better for any new link to assets . After modifying the alias URL, the asset stops working, which is the worst-case scenario. Event If there is potentially risky small malicious, I would prefer issue free version for end users.

Or disable modify alias on edit Asset (UI, API etc.)

@escopecz WDYT?

escopecz commented:
I think we should keep the `/ID:alias` there for BC. It would be really messy to fix that for users and close to impossible if the asset links are in sent emails. I'd go with this solution.

We can also support the UUID but that would be an addition to the previous links.

RCheesley committed:
Merge branch '7.x' into feature/not-possible-to-download-assets-using-only-IDs

RCheesley commented:
Thanks for the input @kuzmany and @escopecz - can we get a +1 on the review if you're happy from the code perspective for this to be merged? Cheers!

escopecz commented on app/bundles/AssetBundle/Controller/PublicController.php:
```diff
@@ -1,92 +1,137 @@
+    ): Response {
+        try {
+            $entity = $model->getRepository()->findByIdAndAlias($slug);
+        } catch (\Exception) {
```
This isn't the use case to catch all exceptions. Please catch only the exception you want to catch. I doubt it's all of them. Code like this is hiding real issues from the logs.

escopecz commented on app/bundles/AssetBundle/Controller/PublicController.php:
```diff
@@ -1,92 +1,137 @@
+            $entity->setUploadDir($parametersHelper->get('upload_dir'));
+            $contents = $entity->getFileContents();
+            $model->trackDownload($entity, $request);
+        } catch (\Exception) {
```
Same here,

escopecz reviewed (changes requested):
The code looks good. Just 1 issue on 2 places.

shinde-rahul committed:
Merge branch '7.x' into feature/not-possible-to-download-assets-using-only-IDs

shinde-rahul committed:
Addressed review comment

Bastian2718 reviewed (approved):
The Bug where a wrong assetalias leads into a download is fixed.
Looks good to me.

escopecz commented on app/bundles/AssetBundle/Controller/PublicController.php:
```diff
@@ -1,92 +1,140 @@
+            $entity->setUploadDir($parametersHelper->get('upload_dir'));
+            $contents = $entity->getFileContents();
+            $model->trackDownload($entity, $request);
+        } catch (FileNotFoundException|ORMException|\Exception) {
```
The change won't help when we are still catching the generic `\Exception`

shinde-rahul replied:
The `\Mautic\AssetBundle\Model\AssetModel::trackDownload` throws the `\Exception`, check these lines 
https://github.com/mautic/mautic/blob/52720645b0e4b6331e722b982f9be5ac74419850/app/bundles/AssetBundle/Model/AssetModel.php#L110-L116
https://github.com/mautic/mautic/blob/52720645b0e4b6331e722b982f9be5ac74419850/app/bundles/AssetBundle/Model/AssetModel.php#L274-L283

I can take this up or I can create an issue for someone else to address as part of a cleanup task — along with [@throws \Exception](https://github.com/search?q=repo%3Amautic%2Fmautic+%22%40throws+%5CException%22&type=code) usages across the codebase.
If you can help me assess the worth of this effort, I’ll proceed accordingly.

escopecz replied:
That try-catch is pointless indeed. Exceptions are already showed in the dev environment and logged in production. Feel free to change it in this PR. Or another one. But it doesn't mean it throws an `\Exception`. Only that it catches it. Don't trust method annotations. There are many in Mautic that are not true.

In this case I think you should catch only the `FileNotFoundException` exception as you don't want to log 404 for an ORM exception. And if an ORM exception happens, it usually means the Entity Manager is closed and you won't be able to store the Download entity anyway.

shinde-rahul committed:
Update PublicController.php

shinde-rahul committed:
Merge branch '7.x' into feature/not-possible-to-download-assets-using-only-IDs

escopecz commented on app/bundles/AssetBundle/Controller/PublicController.php:
```diff
@@ -1,92 +1,140 @@
+        AssetModel $model,
+        string $slug,
+    ): Response {
+        try {
+            $entity = $model->getRepository()->findByIdAndAlias($slug);
+        } catch (NonUniqueResultException|EntityNotFoundException|\InvalidArgumentException) {
+            return $this->notFound();
+        }
+
+        return $this->createAssetResponse($request, $parametersHelper, $model, $entity);
```
Nitpick, but this would read better:
```suggestion
        try {
            $entity = $model->getRepository()->findByIdAndAlias($slug);

            return $this->createAssetResponse($request, $parametersHelper, $model, $entity);
        } catch (NonUniqueResultException|EntityNotFoundException|\InvalidArgumentException) {
            return $this->notFound();
        }
```

escopecz commented on app/bundles/AssetBundle/Controller/PublicController.php:
```diff
@@ -1,92 +1,140 @@
+     */
+    private function createAssetResponse(Request $request, CoreParametersHelper $parametersHelper, AssetModel $model, ?Asset $entity): Response
+    {
+        if (!$this->isAccessAllowed($entity)) {
+            $model->trackDownload($entity, $request, 401);
+            $response = $this->accessDenied();
+        } elseif ($entity->isRemote()) {
+            $response = $this->remoteRedirectResponse($model, $entity, $request);
+        } else {
+            $response = $this->localDownloadResponse($model, $entity, $request, $parametersHelper);
+        }
+
+        return $response;
```
Again, nitpick, but this reads better by following the "return early" rule:
```suggestion
        if (!$this->isAccessAllowed($entity)) {
            $model->trackDownload($entity, $request, 401);

            return $this->accessDenied();
        } elseif ($entity->isRemote()) {
            return $this->remoteRedirectResponse($model, $entity, $request);
        }

        return $this->localDownloadResponse($model, $entity, $request, $parametersHelper);
```

escopecz reviewed (approved):
I found 2 more nitpick suggestions. No need to implement. Good to go from the code review perspective 👍

escopecz commented:
@kuzmany that is a good point!

I guess I didn't understand your comment before. You are saying to keep the current route as is and for new links use UUID? I think that is better.

shinde-rahul committed:
Merge branch '7.x' into feature/not-possible-to-download-assets-using-only-IDs

shinde-rahul committed:
Merge branch '7.x' into feature/not-possible-to-download-assets-using-only-IDs

kuzmany commented:
@shinde-rahul It seems good to me 👍

escopecz committed:
Merge branch '7.x' into feature/not-possible-to-download-assets-using-only-IDs

shinde-rahul commented:
@escopecz, the code for the expected behaviour has not been pushed yet, as I was waiting for the inputs. Let's revert the PR I'll push the changes.
